core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -1731,6 +1764,57 @@ object Parser {
     */
   def end: Parser0[Unit] = Impl.EndParser
 
+  /** return the input string we are parsing.
+    */
+  def input: Parser0[String] = Impl.Input
+
+  /** return the position the parser is at
+    */
+  def careting: Parser0[Caret] =
+    (input, index).tupled map { case (in, ix) =>
+      val locmap = LocationMap(in)
```

Comment:
what about making `careting` the fundamental thing instead of using input, then adding to `State` a `lazy val locationMap: LocationMap = LocationMap(str)`:

https://github.com/typelevel/cats-parse/blob/8c9a27143abc30bc7f8000feb25d75e9a3a10906/core/shared/src/main/scala/cats/parse/Parser.scala#L1880

then you only compute the LocationMap once per parse and only if needed by a parser?

Reply:
Thanks! Now I have `Impl.WithCaret` instead and a `val locmap` in `State`.